# ResMAG

[![Snakemake](https://img.shields.io/badge/snakemake-≥6.3.0-brightgreen.svg)](https://snakemake.github.io)
[![GitHub actions status](https://github.com/<owner>/<repo>/workflows/Tests/badge.svg?branch=main)](https://github.com/<owner>/<repo>/actions?query=branch%3Amain+workflow%3ATests)


ResMAG is a state-of-the-art and user-friendly Snakemake workflow designed for the analysis of metagenomic data. It integrates multiple bioinformatics tools and algorithms to facilitate key steps in metagenome analysis, including bin refinement, taxonomic classification of MAGs,metagenome-assembled genome (MAG) reconstruction and, identification of antibiotic resistance genes.<br />

**Key Features:**<br />

**Binning Techniques**: Employ a collection of five state-of-the-art binning tools to partition metagenomic contigs into individual bins, allowing for comprehensive and accurate analysis.<br />

**MAG Reconstruction**: Utilize cutting-edge algorithms to reconstruct high-quality metagenome-assembled genomes (MAGs) from sequencing data.<br />

**Taxonomic Classification**: Apply advanced taxonomic classification methods to assign taxonomic labels to MAGs and identify the microbial community composition within the metagenomic samples.<br />

**Antibiotic Resistance Gene Identification**: Perform in-depth analysis to detect and characterize antibiotic resistance genes within the metagenomic data, providing valuable insights into antimicrobial resistance profiles.<br />

**Performance Refinement**: Continuously optimize the pipeline by incorporating the latest advancements in metagenomics research, ensuring the highest accuracy and efficiency in metagenomic data analysis.<br />


## Usage

The usage of this workflow is described in the [Snakemake Workflow Catalog](https://snakemake.github.io/snakemake-workflow-catalog/?usage=<owner>%2F<repo>).

If you use this workflow in a paper, don't forget to give credits to the authors by citing the URL of this (original) <repo>sitory and its DOI (see above).

```snakemake --use-conda --cores all --rerun-incomplete```


## Output

## Contributing

**Bug report**

**Feature request**

## License

ResMAG is released under the [BSD-2 Clause](https://www.open-xchange.com/hubfs/2_Clause_BSD_License.pdf?hsLang=en). Please review the license file for more details.

## Contact Information

For any questions, or feedback, please contact the project maintainer at katharina.block@uk-essen.de or josefa.welling@uk-essen.de. We appreciate your input and support in using and improving ResMAG.

## Acknowledgements

We would like to express our gratitude towards Alexander Thomas, Johannes Köster, Ann-Kathrin Brüggemann and the IKIM who have contributed to the development and testing of ResMAG. Their valuable insights and feedback have been helpful throughout the creation of the workflow.
IKIM, Johannes Köster, Alexander Thomas, ?

## References

**Tools**

[Fastp](https://snakemake.github.io)
[Rosella](https://snakemake.github.io)
[MetaBAT2](https://snakemake.github.io)
[MetaBinner](https://snakemake.github.io)
[VAMB](https://snakemake.github.io)
[MetaCoAG](https://snakemake.github.io)
[DASTool](https://snakemake.github.io)
[Snakemake](https://snakemake.github.io)
[Snakemake](https://snakemake.github.io)
[Snakemake](https://snakemake.github.io)
[Snakemake](https://snakemake.github.io)
[Snakemake](https://snakemake.github.io)
[Snakemake](https://snakemake.github.io)
[Snakemake](https://snakemake.github.io)
[Snakemake](https://snakemake.github.io)


**Literature**
[Not here yet](https://www.lipsum.com/feed/html)